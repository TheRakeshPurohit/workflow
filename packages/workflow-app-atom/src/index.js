import { platform } from 'shared/apps';
import execa from 'execa';

const timeout = n => new Promise(resolve => setTimeout(resolve, n));

export const Atom = {
  type: 'app',
  name: 'Atom',
  xClass: 'Atom',

  open: platform({
    'osx-default': osxOpen,
    'linux-*': linuxOpen,
  }),

  async update({ file, position, windowId }, { send, run }) {
    await run(
      (position, windowId) => {
        const Atom = Application('Atom');
        const window = Atom.windows.byId(windowId);
        window.bounds = position;

        return window.id();
      },
      position,
      windowId
    );

    if (send) {
      await send({ topic: 'workflow.apply', message: { file } });
    }
  },
};

async function linuxOpen({ file, appId }) {
  return `WORKFLOW_APP_INSTANCE_ID=${appId} atom -dn ${file}`;
}

async function linuxUpdate({ file }, { send }) {
  send({ topic: 'workflow.apply', message: { file } });
}

async function osxOpen({ file, appId, position }, { run }) {
  console.log('atom', position);
  await execa('atom', ['-dn', file], { env: { WORKFLOW_APP_INSTANCE_ID: appId } });
  await timeout(2000);

  const windowId = await run(position => {
    const Atom = Application('Atom');
    Atom.activate();

    const window = Atom.windows[0];
    window.bounds = position;

    let id = 0;
    while (id <= 0) {
      id = window.id();
    }

    return id;
  }, position);

  console.log('Atom', windowId);

  return { windowId };
}

async function osxUpdate({ file, windowId, position }, { send, run }) {
  await run(
    (windowId, position) => {
      const Atom = Application('Atom');
      const window = Atom.windows.byId(windowId);
      window.bounds = position;
    },
    windowId,
    position
  );

  await send({ topic: 'workflow.apply', message: { file } });
}
