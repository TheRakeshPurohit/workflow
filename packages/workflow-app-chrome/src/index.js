/* global Application */
import { platform } from 'shared/apps';

const buildUrl = ({ url, appId }) =>
  `http://localhost:9876/register?data=${encodeURIComponent(
    JSON.stringify({ redirect: url, appId })
  )}`;

export const Chrome = {
  type: 'app',
  name: 'Chrome',
  xClass: 'Google-chrome',
  connected: true,
  params: ['url', 'appId'],
  open: platform({
    'win32-default': openWindows,
    'osx-default': openOsx,
    'linux-*': openLinux,
  }),

  update: platform({
    'osx-default': updateOsx,
    'linux-*': updateLinux,
  }),
};

function openLinux({ url, appId }) {
  return `google-chrome-stable --new-window ${buildUrl({ url, appId })}`;
}

async function openOsx({ url, appId, position }, { run }) {
  console.log('chromw', position);
  const windowId = await run(
    (url, position) => {
      const chrome = Application('Google Chrome');

      chrome.activate();

      const window = chrome.Window().make();
      window.tabs[0].url = url;
      window.bounds = position;

      let id = 0;
      while (id <= 0) {
        id = window.id();
        delay(0.1);
      }
      return id;
    },

    buildUrl({ url, appId }),
    position
  );

  console.log('chrome', windowId);

  return { windowId };
}

async function openWindows({ url, appId, position }, { startOnPositionByWindowClass }) {
  await startOnPositionByWindowClass({
    cmd: 'chrome.exe',
    args: ['--new-window', buildUrl({ url, appId })],
    className: 'Chrome_WidgetWin_1',
    position,
  });
}

async function updateLinux({ url }, { send }) {
  if (send) {
    await send({ topic: 'workflow.apply', message: { url } });
  }
}

async function updateOsx({ url, windowId, position }, { send, run }) {
  await run(
    (windowId, position) => {
      const chrome = Application('Google Chrome');
      const window = chrome.windows.byId(windowId);
      window.bounds = position;
    },
    windowId,
    position
  );

  console.log('sending to chrome');
  if (send) {
    await send({ topic: 'workflow.apply', message: { url } });
  }
}
