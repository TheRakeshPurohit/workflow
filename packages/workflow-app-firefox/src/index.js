/* eslint-disable no-unused-vars */
export const Firefox = {
  type: 'app',
  name: 'Firefox',
  params: ['url'],
  open: async ({ url, position }, context, children) => {
    const { run } = context;

    await run({
      cmd: 'firefox',
      args: ['--new-window', url],
      position,
      className: 'Navigator.Firefox',
    });
  },
};
