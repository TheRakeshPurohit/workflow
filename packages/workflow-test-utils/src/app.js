/* eslint-env jest */
export function testApp(node) {
  const { name } = node;
  test('app should have name of type string', () => {
    expect(name).toBeTruthy();
    expect(typeof name).toEqual('string');
  });

  test(`${name} should have open function`, () => {
    expect(typeof node.open).toEqual('function');
  });

  test(`${name} should have params array`, () => {
    expect(Array.isArray(node.params)).toBeTruthy();
  });

  if (node.xClass) {
    test(`${name}.xClass should be string`, () => {
      expect(typeof node.xClass).toEqual('string');
    });
  }
}
