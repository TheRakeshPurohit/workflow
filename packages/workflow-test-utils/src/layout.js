/* eslint-env jest */
export function testLayout(node) {}
