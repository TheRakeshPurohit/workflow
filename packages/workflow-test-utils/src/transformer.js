/* eslint-env jest */
export function testTransformer({ Transformer }) {
  const instance = new Transformer();

  if (instance.transformBefore) {
    test('transformBefore should a function', () => {
      expect(typeof instance.transformBefore).toEqual('function');
    });

    test('transformBefore should return node on unknown type', async () => {
      const node = { type: 'unknown' };
      expect(await instance.transformBefore(node)).toBe(node);
    });
  }

  if (instance.transformAfter) {
    test('transformAfter should a function', () => {
      expect(typeof instance.transformAfter).toEqual('function');
    });

    test('transformAfter should return node on unknown type', async () => {
      const node = { type: 'unknown' };
      expect(await instance.transformAfter(node)).toBe(node);
    });
  }
}
