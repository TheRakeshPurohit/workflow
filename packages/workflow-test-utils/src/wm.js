/* eslint-env jest */
export function testWm() {}
