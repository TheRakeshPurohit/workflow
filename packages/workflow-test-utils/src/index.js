import { testApp } from './app';
import { testLayout } from './layout';
import { testWm } from './wm';
export * from './transformer';

export function testNode(node) {
  switch (node.type) {
    case 'app':
      return testApp(node);
    case 'layout':
      return testLayout(node);
    case 'wm':
      return testWm(node);
    default:
      throw new Error('Unknown node type ' + node.type);
  }
}
