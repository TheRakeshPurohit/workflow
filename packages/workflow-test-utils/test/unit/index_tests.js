import { testApp } from '../../src/app';
import { testLayout } from '../../src/layout';
import { testWm } from '../../src/wm';
import { testNode } from '../../src';

jest.mock('../../src/app', () => ({ testApp: jest.fn() }));
jest.mock('../../src/layout', () => ({ testLayout: jest.fn() }));
jest.mock('../../src/wm', () => ({ testWm: jest.fn() }));

test('testNode(unknown) should throw error', () => {
  const app = { type: 'unknown' };

  expect(() => testNode(app)).toThrowErrorMatchingSnapshot();
});

test('testNode(app) should call testApp', () => {
  const app = { type: 'app' };

  testNode(app);

  expect(testApp).toBeCalledWith(app);
});

test('testNode(layout) should call testLayout', () => {
  const layout = { type: 'layout' };

  testNode(layout);

  expect(testLayout).toBeCalledWith(layout);
});

test('testNode(wm) should call testWm', () => {
  const wm = { type: 'wm' };

  testNode(wm);

  expect(testWm).toBeCalledWith(wm);
});
