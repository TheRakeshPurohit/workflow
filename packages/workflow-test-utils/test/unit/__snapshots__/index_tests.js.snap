// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testNode(unknown) should throw error 1`] = `"Unknown node type unknown"`;
